Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar, which is a grammar that has a single
+nonterminal symbol on the left-hand side of each rule.
+-/
+
+/-- Rule that rewrites a single nonterminal to any list of symbols. -/
+structure ContextFreeRule (T : Type*) (N : Type*) where
+  /-- Input nonterminal a.k.a. left-hand side -/
+  input : N
+  /-- Output string a.k.a. right-hand side -/
+  output : List (Symbol T N)
+
+/-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
+structure ContextFreeGrammar (T : Type*) where
+  /-- Type of nonterminals -/
+  NT : Type*
+  /-- Initial nonterminal -/
+  initial : NT
+  /-- Rewrite rules -/
+  rules : List (ContextFreeRule T NT)
+
+variable {T : Type*}
+
+/-- One application of single context-free rule. -/
+inductive ContextFreeRule.RewritesTo {N : Type*} (r : ContextFreeRule T N) :
+    List (Symbol T N) → List (Symbol T N) → Prop
+  /-- The replacement is at the start of the remaining string. -/
+  | head (s : List (Symbol T N)) :
+      r.RewritesTo (Symbol.nonterminal r.input :: s) (r.output ++ s)
+  /-- There is a replacement later in the string. -/
+  | cons (x : Symbol T N) {s₁ s₂ : List (Symbol T N)} (hrs : RewritesTo r s₁ s₂) :
+      r.RewritesTo (x :: s₁) (x :: s₂)
+
+lemma ContextFreeRule.RewritesTo.exists_parts {N : Type _} {r : ContextFreeRule T N}
+    {u v : List (Symbol T N)} (hyp : r.RewritesTo u v) :
+    ∃ p : List (Symbol T N), ∃ q : List (Symbol T N),
+      u = p ++ [Symbol.nonterminal r.input] ++ q ∧ v = p ++ r.output ++ q := by
+  induction hyp with
+  | head xs =>
+    use [], xs
+    simp
+  | cons x _ ih =>
+    rcases ih with ⟨p', q', bef, aft⟩
+    use x :: p', q'
+    simp [bef, aft]
+
+lemma ContextFreeRule.rewritesTo_of_exists_parts {N : Type _}
+    (r : ContextFreeRule T N) (p q : List (Symbol T N)) :
+    r.RewritesTo (p ++ [Symbol.nonterminal r.input] ++ q) (p ++ r.output ++ q) := by
+  induction p with
+  | nil         => exact ContextFreeRule.RewritesTo.head q
+  | cons d l ih => exact ContextFreeRule.RewritesTo.cons d ih
+
+theorem ContextFreeRule.rewritesTo_iff {N : Type _} {r : ContextFreeRule T N}
+    (u v : List (Symbol T N)) :
+    r.RewritesTo u v ↔
+      ∃ p : List (Symbol T N), ∃ q : List (Symbol T N),
+        u = p ++ [Symbol.nonterminal r.input] ++ q ∧ v = p ++ r.output ++ q := by
+  constructor
+  · apply ContextFreeRule.RewritesTo.exists_parts
+  · rintro ⟨p, q, bef, aft⟩
+    rw [bef, aft]
+    apply ContextFreeRule.rewritesTo_of_exists_parts
+
+namespace ContextFreeGrammar
+
+/-- If `g` is a context-free grammar, `g.Produces u v` means that one step of context-free
```

Comment:
Please write this style of docstring for `Derives` and `RewritesTo` as well.

Reply:
Will do!